### PR TITLE
Envelope generator

### PR DESCRIPTION
**Pull request by jpcima**
This rewrites EG in such a way that:
- decay operates more similar to other products
- no exp/log calls in the process function

Fails the tests currently.
It may be the case that the release segment takes longer to reach 0, keeping the voice alive longer.
Release is considered finished when EG < virtuallyZero (-86dB). This threshold could be set higher.

Below are Cakewalk, Sfizz, ARIA (in that order)
Capture du 2020-03-25 19-58-21

### CODE REVIEW

**Comment by jpcima** [comment]
Below is a comparison to Dimension, which has different EG that Dropzone.
It seems to fix the release time.

It's observable that Dimension has a slightly longer smooth duration.
Current sfizz is between the two of these, it is adjustable by the parameter of the exponential rate.

Capture du 2020-03-25 22-02-15

**Comment by jpcima** [comment]
After slight adjusting.
Exponential factor=exp(-8/(t*Fs))

Cakewalk/Sfizz/ARIA
Capture du 2020-03-25 22-12-14

**Comment by jpcima** [comment]
The tool is a Jack program that permits to capture the exact EG curves from Dimension with very little manual work.
This can be used in order to build the test set.

**Comment by jpcima** [comment]
@paulfd Do you know an appropriate technique to compare the similarity of 2 curves overall?
The maximum absolute error does not work well.

---

For instance here are Sfizz vs Dim versions of the "Attack" envelope test.
They appear greatly similar at the first look.

Capture du 2020-03-27 12-58-48

When zooming into the release segment, it's observed there are large absolute errors locally, >0.1 in some cases.
Cakewalk looks like a reverse S-shape, suggests a different kind of smoothing. (maybe 2nd-order?)

Capture du 2020-03-27 12-59-58

**Comment by paulfd** [comment]
Max absolute error will greatly favor outliers I would say. The mean squared/absolute error on the other hand does not tell you much about how the curves actually behave w.r.t. each other. I think we'll need to make it very specific for each test case: find the attack/sustain/release/... ranges, and compute a mean squared/absolute error here.

Between mean-squared and absolute, or any other penalty, I guess there are not really a best choice and it kinda depends on what will sound OK. I'd go with mean-squared or root-mean-squared due to habit but as I said, I'm not super clear on the psycho-acoustic part of what the best penalty would be for things to sound "the same".

**Comment by paulfd** [comment]
You could also penalize the derivative if you want to capture smoothness effects.

**Comment by jpcima** [comment]
Attached: some EG recordings from Dimension to match current unit tests
[eg-recordings.tar.gz](https://github.com/sfztools/sfizz/files/4400012/eg-recordings.tar.gz)

File: src/sfizz/ADSREnvelope.cpp [hunk 1]
```diff
@@ -100,110 +99,92 @@ Type ADSREnvelope<Type>::getNextValue() noexcept
...
+        case State::Decay:
+            while (count < size && (currentValue *= decayRate) > sustain)
+                output[count++] = currentValue;
+            if (currentValue <= sustain) {
+                currentValue = sustain;
+                currentState = State::Sustain;
+            }
+            break;
+        case State::Sustain:
+            if (!shouldRelease && freeRunning) {
```
Comment (paulfd) [inline src/sfizz/ADSREnvelope.cpp:158]: Is the `!shouldRelease` check necessary here?

File: src/sfizz/ADSREnvelope.cpp [hunk 2]
```diff
@@ -100,110 +99,92 @@ Type ADSREnvelope<Type>::getNextValue() noexcept
...
+                shouldRelease = true;
+                break;
+            }
+            count = size;
+            currentValue = sustain;
+            sfz::fill(output.first(count), currentValue);
+            break;
+        case State::Release:
+            while (count < size && (currentValue *= releaseRate) > config::virtuallyZero)
```
Comment (paulfd) [inline src/sfizz/ADSREnvelope.cpp:167]: So you're saying this could be done earlier? I'm not clear on the "I don't hear anything" level to be honest :)

**Review by paulfd — APPROVED** [review]
I will update the test when #142 is merged. The rebasing seems simpler in this order !

File: src/sfizz/ADSREnvelope.cpp (hunk 2 above)
Comment (jpcima) [reply src/sfizz/ADSREnvelope.cpp:167]: Usually -60 dB, -80 dB sometimes

**Comment by paulfd** [comment]
Btw I'm tempted to remove completely the `getNextValue()` API, since it's never used, I think your block version is also a bit "per-sample" but has small loops and should always perform better anyway. 1 less code path to maintain.

**Pull request merged**